ksql-engine/src/main/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNode.java

```diff
@@ -244,12 +244,71 @@ private static void createSinkTopic(
         : Collections.emptyMap();
 
     kafkaTopicClient.createTopic(kafkaTopicName,
-                                 numberOfPartitions,
-                                 numberOfReplications,
-                                 config
+        numberOfPartitions,
+        numberOfReplications,
+        config
     );
   }
 
+  private static SourceTopicProperties getSourceTopicProperties(
+      final String kafkaTopicName,
+      final Map<String, Object> sinkProperties,
+      final KafkaTopicClient kafkaTopicClient,
+      final KsqlConfig ksqlConfig
+  ) {
+    final Map ksqlProperties = ksqlConfig.values();
+    if ((ksqlProperties.containsKey(KsqlConfig.SINK_NUMBER_OF_PARTITIONS_PROPERTY)
+          && ksqlProperties.get(KsqlConfig.SINK_NUMBER_OF_PARTITIONS_PROPERTY) != null)
+        || (ksqlProperties.containsKey(KsqlConfig.SINK_NUMBER_OF_REPLICAS_PROPERTY)
+          && ksqlProperties.get(KsqlConfig.SINK_NUMBER_OF_REPLICAS_PROPERTY) != null)) {
+      return getSinkTopicPropertiesLegacyWay(sinkProperties, ksqlConfig);
+    }
+    // Don't request topic properties from Kafka if both are set in WITH clause.
+    if (sinkProperties.containsKey(KsqlConfig.SINK_NUMBER_OF_PARTITIONS_PROPERTY)
```

Comment:
defensive programming: consider favouring `sinkProperties.get(blah) != null` over `sinkProperties.containsKey(blah)` as this will avoid NPEs should the map contain null values.